File: src/public/PaywallPresentationRequestStatus.ts
import { Experiment } from './Experiment';

export enum PaywallPresentationRequestStatusType {
  presentation = 'presentation',
  noPresentation = 'noPresentation',
  timeout = 'timeout',
}

export class PaywallPresentationRequestStatus {
  type: PaywallPresentationRequestStatusType;

  private constructor(type: PaywallPresentationRequestStatusType) {
    this.type = type;
  }

  static fromJson(json: {
    [key: string]: any;
  }): PaywallPresentationRequestStatus {
    switch (json.status) {
      case 'presentation':
        return new PaywallPresentationRequestStatus(
          PaywallPresentationRequestStatusType.presentation
        );
      case 'noPresentation':
        return new PaywallPresentationRequestStatus(
          PaywallPresentationRequestStatusType.noPresentation
        );
      case 'timeout':
        return new PaywallPresentationRequestStatus(
          PaywallPresentationRequestStatusType.timeout
        );
      default:
        throw new Error('Invalid PaywallPresentationRequestStatus type');
    }
  }
}

export enum PaywallPresentationRequestStatusReasonType {
  debuggerPresented = 'debuggerPresented',
  paywallAlreadyPresented = 'paywallAlreadyPresented',
  userIsSubscribed = 'userIsSubscribed',
  holdout = 'holdout',
  noAudienceMatch = 'noAudienceMatch',
  placementNotFound = 'placementNotFound',
  noPaywallViewController = 'noPaywallViewController',
  noPresenter = 'noPresenter',
  noConfig = 'noConfig',
  subscriptionStatusTimeout = 'subscriptionStatusTimeout',
}

export class PaywallPresentationRequestStatusReason {
  type: PaywallPresentationRequestStatusReasonType;
  experiment?: Experiment;

  private constructor(
    type: PaywallPresentationRequestStatusReasonType,
    experiment?: Experiment
  ) {
    this.type = type;
    this.experiment = experiment;
  }

  static fromJson(json: {
    [key: string]: any;
  }): PaywallPresentationRequestStatusReason {
    switch (json.reason) {
      case 'debuggerPresented':
        return new PaywallPresentationRequestStatusReason(
          PaywallPresentationRequestStatusReasonType.debuggerPresented
        );
      case 'paywallAlreadyPresented':
        return new PaywallPresentationRequestStatusReason(
          PaywallPresentationRequestStatusReasonType.paywallAlreadyPresented
        );
      case 'userIsSubscribed':
        return new PaywallPresentationRequestStatusReason(
          PaywallPresentationRequestStatusReasonType.userIsSubscribed
        );
      case 'holdout':
        return new PaywallPresentationRequestStatusReason(
          PaywallPresentationRequestStatusReasonType.holdout,
          Experiment.fromJson(json.experiment)
        );
      case 'noAudienceMatch':
        return new PaywallPresentationRequestStatusReason(
          PaywallPresentationRequestStatusReasonType.noAudienceMatch
        );
      case 'placementNotFound':
        return new PaywallPresentationRequestStatusReason(
          PaywallPresentationRequestStatusReasonType.placementNotFound
        );
      case 'noPaywallViewController':
        return new PaywallPresentationRequestStatusReason(
          PaywallPresentationRequestStatusReasonType.noPaywallViewController
        );
      case 'noPresenter':
        return new PaywallPresentationRequestStatusReason(
          PaywallPresentationRequestStatusReasonType.noPresenter
        );
      case 'noConfig':
        return new PaywallPresentationRequestStatusReason(
          PaywallPresentationRequestStatusReasonType.noConfig
        );
      case 'subscriptionStatusTimeout':
        return new PaywallPresentationRequestStatusReason(
          PaywallPresentationRequestStatusReasonType.subscriptionStatusTimeout
        );
      default:
        throw new Error('Invalid PaywallPresentationRequestStatusReason type');
    }
  }
}

File: android/src/main/java/com/superwallreactnative/models/PaywallPresentationRequestStatus.kt
package com.superwallreactnative.models

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReadableMap
import com.superwall.sdk.paywall.presentation.internal.PaywallPresentationRequestStatus
import com.superwall.sdk.paywall.presentation.internal.PaywallPresentationRequestStatusReason

fun PaywallPresentationRequestStatus.toJson(): ReadableMap {
  val map = Arguments.createMap()
  val statusValue = when (this) {
    PaywallPresentationRequestStatus.Presentation -> "presentation"
    PaywallPresentationRequestStatus.NoPresentation -> "noPresentation"
    PaywallPresentationRequestStatus.Timeout -> "timeout"
  }
  map.putString("status", statusValue)
  return map
}

fun PaywallPresentationRequestStatusReason.toJson(): ReadableMap {
  val map = Arguments.createMap()
  when (this) {
    is PaywallPresentationRequestStatusReason.DebuggerPresented -> map.putString("reason", "debuggerPresented")
    is PaywallPresentationRequestStatusReason.PaywallAlreadyPresented -> map.putString("reason", "paywallAlreadyPresented")
    is PaywallPresentationRequestStatusReason.Holdout -> {
      map.putString("reason", "holdout")
      map.putMap("experiment", Experiment.toJson(this.experiment))
    }
    is PaywallPresentationRequestStatusReason.NoAudienceMatch -> map.putString("reason", "noAudienceMatch")
    is PaywallPresentationRequestStatusReason.PlacementNotFound -> map.putString("reason", "placementNotFound")
    is PaywallPresentationRequestStatusReason.NoPaywallView -> map.putString("reason", "noPaywallViewController")
    is PaywallPresentationRequestStatusReason.NoPresenter -> map.putString("reason", "noPresenter")
    is PaywallPresentationRequestStatusReason.NoConfig -> map.putString("reason", "noConfig")
    is PaywallPresentationRequestStatusReason.SubscriptionStatusTimeout -> map.putString("reason", "subscriptionStatusTimeout")
  }
  return map
}
